src/syntax/syntax.rkt

```diff
@@ -60,44 +60,136 @@
   [cost  natural-number/c]
-  [type  (hash/c (or/c '* natural-number/c) (list/c (or/c (listof type?) (list/c '* type?)) type?))]
+  [type  (hash/c (or/c '* natural-number/c) (listof (list/c (or/c (listof type?) (list/c '* type?)) type?)))]
   [->c/double (unconstrained-argument-number-> string? string?)]
   [->c/mpfr   (unconstrained-argument-number-> string? string?)]
   [->tex      (unconstrained-argument-number-> string? string?)])
 
 (define (operator-info operator field) (table-ref operators operator field))
 
 (define-syntax-rule (define-operator (operator atypes ...) rtype [key value] ...)
-  (let ([type (hash (length '(atypes ...)) (list '(atypes ...) 'rtype))]
+  (let ([type (hash (length '(atypes ...)) (list (list '(atypes ...) 'rtype)))]
         [args (list (length '(atypes ...)))])
     (table-set! operators 'operator
                 (make-hash (list (cons 'type type) (cons 'args args) (cons 'key value) ...)))))
 
-(define-operator (+ real real) real
-  [fl +] [bf bf+] [cost 40]
+(define (bffmod x mod)
+  (bf- x (bf* mod (bffloor (bf/ x mod)))))
+
+(define (bf-make-rectangular x y)
+  (cons x y))
+
+(define bf-real-part car)
+(define bf-imag-part cdr)
+
+(define (bf-complex-add x y)
+  (cons (bf+ (bf-real-part x) (bf-real-part y)) (bf+ (bf-imag-part x) (bf-imag-part y))))
+
+(define (bf-complex-sub x [y #f])
+  (if y
+      (bf-complex-add x (bf-complex-neg y))
+      (bf-complex-neg x)))
+
+(define (bf-complex-neg x)
+  (cons (bf- (bf-real-part x)) (bf- (bf-imag-part x))))
+
+(define (bf-complex-mult x y)
+  (cons (bf+ (bf* (bf-real-part x) (bf-real-part y)) (bf- (bf* (bf-imag-part x) (bf-imag-part y))))
+        (bf+ (bf* (bf-imag-part x) (bf-real-part y)) (bf* (bf-real-part x) (bf-imag-part y)))))
+
+(define (bf-complex-conjugate x)
+  (cons (bf-real-part x) (bf- (bf-imag-part x))))
+
+(define (bf-complex-sqr x)
+  (bf-complex-mult x x))
+
+(define (bf-complex-exp x)
+  (match-define (cons re im) x)
+  (define scale (bfexp re))
+  (cons (bf* scale (bfcos im)) (bf* scale (bfsin im))))
+
+(define (bf-complex-log x)
+  (match-define (cons re im) x)
+  (define mag (bfhypot re im))
+  (define arg (bfatan2 im re))
+  (cons (bflog mag) arg))
+
+(define (bf-complex-sqrt x)
+  (bf-complex-pow x (cons (bf 0.5) 0.bf)))
+
+(define (bf-complex-pow x n)
+  (bf-complex-exp (bf-complex-mult n (bf-complex-log x))))
+
+(define (bf-complex-div x y)
+  (define numer (bf-complex-mult x (bf-complex-conjugate y)))
+  (define denom (bf-complex-mult y (bf-complex-conjugate y)))
+  (cons (bf/ (bf-real-part numer) (bf-real-part denom)) (bf/ (bf-imag-part numer) (bf-real-part denom))))
+
+(define (make-exact-fun bf-fun bf-complex-fun)
+  (lambda args
+    (match args
+      [(list (? bigfloat?) ...)
+       (apply bf-fun args)]
+      [(list (? pair?) ...)
+       (apply bf-complex-fun args)])))
+
+(require (only-in racket/base [exp e]))
+#;(define exp-for-type
```

Comment:
Let's delete this partial attempt at implementing `exp`—it'll only confuse us later on.

Reply:
deleted